src/m365/entra/commands/app/app-permission-list.ts

```diff
@@ -0,0 +1,264 @@
+        return (servicePrincipal.oauth2PermissionScopes as PermissionScope[])
+          .find(permissionScope => permissionScope.id === permissionId)?.value as string ?? permissionId;
+    }
+    /* c8 ignore next 4 */
+    // permissionType is either 'Scope' or 'Role' but we need a safe default
+    // to avoid building errors. This code will never be reached.
+    return permissionId;
+  }
+
+  private async getServicePrincipal(servicePrincipalInfo: ServicePrincipalInfo, permissionType: string, logger: Logger): Promise<ServicePrincipal | undefined> {
+    if (this.verbose) {
+      await logger.logToStderr(`Retrieving service principal ${servicePrincipalInfo.appId}`);
+    }
+
+    const requestOptions: CliRequestOptions = {
+      url: `${this.resource}/v1.0/servicePrincipals?$filter=appId eq '${servicePrincipalInfo.appId}'&$select=appId,id,displayName`,
+      headers: {
+        accept: 'application/json;odata.metadata=none'
+      },
+      responseType: 'json'
+    };
+
+    const response = await request.get<{ value: ServicePrincipal[] } | ServicePrincipal>(requestOptions);
+
+    if ((servicePrincipalInfo.id) ||
+      (servicePrincipalInfo.appId && (response as { value: ServicePrincipal[] }).value.length === 0)) {
+      return undefined;
+    }
+
+    const servicePrincipal = (response as { value: ServicePrincipal[] }).value[0];
+
+    if (this.verbose) {
+      await logger.logToStderr(`Retrieving permissions for service principal ${servicePrincipal.id}...`);
+    }
+
+    const oauth2PermissionScopesRequestOptions: CliRequestOptions = {
+      url: `${this.resource}/v1.0/servicePrincipals/${servicePrincipal.id}/oauth2PermissionScopes`,
+      headers: {
+        accept: 'application/json;odata.metadata=none'
+      },
+      responseType: 'json'
+    };
+    const appRolesRequestOptions: CliRequestOptions = {
+      url: `${this.resource}/v1.0/servicePrincipals/${servicePrincipal.id}/appRoles`,
+      headers: {
+        accept: 'application/json;odata.metadata=none'
+      },
+      responseType: 'json'
+    };
+
+    let permissions: any;
+
+    switch (permissionType) {
+      case 'all':
+        permissions = await Promise.all([
+          request.get<{ value: PermissionScope[] }>(oauth2PermissionScopesRequestOptions),
+          request.get<{ value: AppRole[] }>(appRolesRequestOptions)
+        ]);
+
+        servicePrincipal.oauth2PermissionScopes = permissions[0].value as PermissionScope[];
+        servicePrincipal.appRoles = permissions[1].value as AppRole[];
+        break;
+
+      case 'delegated':
+        permissions = await request.get<{ value: PermissionScope[] }>(oauth2PermissionScopesRequestOptions);
+        servicePrincipal.oauth2PermissionScopes = permissions.value as PermissionScope[];
+        break;
+
+      case 'application':
+        permissions = await request.get<{ value: AppRole[] }>(appRolesRequestOptions);
+        servicePrincipal.appRoles = permissions.value as AppRole[];
+        break;
+    }
```

Comment:
What do you think of something like this?

```suggestion
 if(permissionType === 'all' || permissionType === 'delegated') {
      permissions = await request.get<{ value: PermissionScope[] }>(oauth2PermissionScopesRequestOptions);
      servicePrincipal.oauth2PermissionScopes = permissions.value as PermissionScope[];
    }
    
    if (permissionType === 'all' || permissionType === 'application') {
      permissions = await request.get<{ value: AppRole[] }>(appRolesRequestOptions);
      servicePrincipal.appRoles = permissions.value as AppRole[];
    }
```

isn't it like smaller and less duplicated? or maybe there is something I am not seeing 🙂

Reply:
It's smaller and works, but a little less practical since it doesn't use `Promise.all` when all permissions need to be collected